src/vm/ast/parser/mod.rs

```diff
@@ -180,11 +180,15 @@ pub fn lex(input: &str) -> ParseResult<Vec<(LexItem, u32, u32)>> {
                         Ok(LexItem::Whitespace)
                     },
                     TokenType::Comma => {
-                        context = LexContext::ExpectNothing;
+                        // comma should not be followed directly by an item,
+                        //  e.g., {a: 1,b: 2} should not be legal
```

Comment:
Why not?

Reply:
I take it that we need whitespace because `1,b` is a valid identifier?

Reply:
while `1,b` would in almost no contexts be a valid identifier (and `a,b` would also rarely be a valid identifier), the behavior here is just enforced to match enforcement behavior of `:`.

Reply:
> a,b would also rarely be a valid identifier

Is `a,b` ever a valid tuple identifier? Tuple identifiers are resolved by `parse_eval_bindings`, which expects tuple identifiers to be atoms originating from `TokenType::Variable` lexed according to `CLARITY_NAME_REGEX`:

https://github.com/blockstack/stacks-blockchain/blob/5aa407ea524002e4c1625c9c36bbeb47d65b240f/src/vm/ast/parser/mod.rs#L89

Apparently such identifiers can only contain alphanumeric characters and a few special characters excluding comma and colon. Or am I missing something?

PS: `TokenType::Variable` is a misnomer - it should be renamed to `TokenType::Symbol`.

Reply:
That's correct: `a,b` is never a valid identifier in Clarity. However, character sequences without spaces are handled in a wide variety of ways in LISPs (for example, `a:b` would be a valid identifier in many variants, which is why we require a space to eliminate any would be confusion there).

Reply:
We're already loosing the Lisp developers by having the decorative colons - the space won't make much of a difference. More of concern should be the implications for the evolution of Clarity. 

The tuple colons does not come for free. Without mandatory whitespace after the colon, it will be complicated to later allow colons in identifiers/symbols. But even with the space, parsing symbols with colons will no longer be simple, suddenly requiring backtracking or multiple character look-ahead. So this would be a good time to decide whether to potentially allow colons in symbols, and if not, drop the required whitespace after the colon.

Reply:
My concern is confusion over whether `a:b` is two identifiers or a single identifier -- and that confusion is independent of whether or not `:` is a legal character in identifiers within Clarity. Languages do not exist in a vacuum, and in the majority of Lisp variants, `a:b` is either an identifier or an illegal name. Regardless of whether or not the language is designed to appeal to existing Lisp developers, parsing `a:b` as two identifiers is surprising and potentially confusing, and I see no advantage to doing so.

Reply:
> parsing a:b as two identifiers is surprising and potentially confusing, and I see no advantage to doing so.

You don't think throwing an error for `{ key:value }` is equally surprising and confusing? Most developers are likely to add a space after the colon anyway. It would be more robust to *not* make the parser err on such insignificant formatting, leaving the syntax recommendations to a linter.

Reply:
For REPL-driven development, robust parsing makes the difference between frustration or delight. In the REPL, developers are apt to take shortcuts for expediency, such as ignoring formatting. When evaluating expressions in the REPL, developers will be frustrated by a strict parser that balks on insignificant whitespace and punctuation.

Reply:
> You don't think throwing an error for { key:value } is equally surprising and confusing? Most developers are likely to add a space after the colon anyway. It would be more robust to not make the parser err on such insignificant formatting, leaving the syntax recommendations to a linter.

No, I do not. An error message from the parser at least makes it clear exactly which expression has errored and what the cause of the error is. To discover whether or not `a:b` parses as a single identifier or two requires consulting a language specification.